File: Content/Script/GameFlow/Services/GameFlowStateService.lua
local GameFlowStateService = UnrealClass("DogFight.Services.GameServiceBase")

function GameFlowStateService:StartupScript(ServiceName)
    self.Super.StartupScript(self, ServiceName)

    -- Setup variable
    self.DefaultGameFlowStateCreateArgumentPath = "GameFlow.GameFlowState.GameFlowStateCreateArgument"
    self.GameFlowStateMachineBasePath = "GameFlow.GameFlowStateMachine."
    self.GameFlowStateBasePath = "GameFlow.GameFlowState."
end

function GameFlowStateService:GetGameFlowStateCreateArgument(Instigator)
    local Class = UE.UGameFlowStateCreateArgument.StaticClass()
    local NewInstance = NewObject(Class, Instigator, nil, self.DefaultGameFlowStateCreateArgumentPath)
    return NewInstance
end

function GameFlowStateService:CreateStateMachine(InPath, Instigator)
    local Class = UE.UGameFlowStateMachine.StaticClass()
    local NewInstance = NewObject(Class, Instigator, nil, self.GameFlowStateMachineBasePath .. InPath)
    return NewInstance
end

return GameFlowStateService

File: Source/GameFlow/Private/GameService/GameFlowStateService.cpp
#include "GameService/GameFlowStateService.h"
#include "GameFlowState/GameFlowStateBase.h"
#include "GameFlowState/GameFlowStateCreateArgument.h"

UGameFlowStateBase* UGameFlowStateService::CreateGameFlowState(UGameFlowStateCreateArgument* InArgument)
{
	if (!IsValid(InArgument))
	{
		return nullptr;
	}

	auto NewInstance = NewObject<UGameFlowStateBase>(InArgument->Instigator);
	if (NewInstance)
	{
		NewInstance->SetCreateArgument(InArgument);
	}
	return NewInstance;
}
